File: python/Stack/150. Evaluate Reverse Polish Notation.py
'''
Time: O(n)
Space: O(n)
'''

class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        stack = []
        for o in tokens:
            if o == '+':
                stack.append(stack.pop() + stack.pop())
            elif o == '*':
                stack.append(stack.pop() * stack.pop())
            elif o =='-':
                num2, num1 = stack.pop(), stack.pop()
                stack.append(num1 - num2)
            elif o == '/':
                num2, num1 = stack.pop(), stack.pop()
                stack.append(int(num1 / num2))
            else:
                stack.append(int(o))
        return stack[0]

File: javascript/Stack/150. Evaluate Reverse Polish Notation.js
/**
 * @param {string[]} tokens
 * @return {number}
 */
var evalRPN = function(tokens) {
    const stack = [];
    for (let o of tokens) {
        switch (o) {
            case '+':
                stack.push(stack.pop() + stack.pop());
                break;
            case '*':
                stack.push(stack.pop() * stack.pop());
                break;
            case '-':
                let num2= stack.pop(), num1 = stack.pop();
                stack.push(num1 - num2);
                break;
            case '/':
                let n2 = stack.pop(), n1 = stack.pop();
                stack.push(n1 / n2 > 0 ? Math.floor(n1 / n2) : Math.ceil(n1 / n2));
                break;
            default:
                stack.push(Number(o));
        }
    }
    return stack[0];
};